### ISSUE TITLE
[no-namespace] Are merged namespaces still discouraged?

### ISSUE THREAD

**Issue by thw0rted** [typescript-eslint/typescript-eslint#2907]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
    "extends": [
        "eslint:recommended",
        "plugin:@typescript-eslint/recommended"
    ],
}
```

```TS
// One.ts
export namespace One {
  export interface Options {
    // ...
  }
}

export class One {
  foo(opts: One.Options) { }
  bar(opts: One.Options} { }
}


// consumer.ts
import { One } from "./One";
import { Two } from "./Two";
// etc

const opt1: One.Options = { x: 1, y: true };
const one = new One();
one.foo(opt1);

const opt2: Two.Options = { a: 0, b: "hi" };
const two = new Two();
two.baz(opt2);
```

**Discussion**

(I realize that this is more open-ended than a bug report, but I wasn't sure where else to post it -- not exactly appropriate for StackOverflow, and the "discussion" tab in GitHub isn't enabled for this project.)

Is the above code considered invalid or an anti-pattern?  I have a large number of related classes that all use different interfaces to describe common method and constructor arguments.  At one point, it was considered ergonomic to use merging to attach the interfaces to the classes they belong to, but under the current default rule-set, the linter flags this usage as an error.

I could refactor the interfaces to be top-level exports, but that comes with a tradeoff.  If I call the interface "Options", consumers that use more than one of these classes wind up having to rename the import, i.e `import {One, Options as OneOptions} from "./One"`.  It also makes multiple star-exports impossible, i.e. `export * from "./One"; export * from "./Two";`.   The other solution is to give each interface a unique name (`OneOptions`, `TwoOptions`, etc) but that gets ugly with e.g. `MyVeryLongClassNameOptions`.

I read https://github.com/typescript-eslint/typescript-eslint/issues/423 and https://github.com/typescript-eslint/typescript-eslint/issues/324 to try to understand the spirit of the error, but they don't address this kind of merging.  I certainly agree that [needless namespacing](https://www.typescriptlang.org/docs/handbook/namespaces-and-modules.html#needless-namespacing) adds extra levels of indirection with no value, but I don't think my use case falls into that category.  I've tried to find best-practice guidance in the TS manual but I don't think there's anything there.

I could disable the rule, but I do want to flag the "needless" type of namespacing as an error, so I'm hoping to come up with consistent, well-reasoned guidance that can be enforced by the linter.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.11.0` |
| `@typescript-eslint/parser`        | `4.11.0` |
| `TypeScript`                       | `4.1.3` |
| `ESLint`                           | `7.15.0` |

**Comment by bradzacher** [comment]
What your codebase does is up to you!

The most common use for namespaces is to create psuedo-modules. That usecase is now obsolete with ES6 module - thus it is a definite anti-pattern.

Merging type-only namespaces into classes isn't a terrible way to take care of the naming clash problem. However it does come at a cost of always requiring you to import the class - you can never import just the type.

There are trade-offs to every approach. If you have a modern codebase that is very heavy on classes, then it might be fine to just disable the rule and use namespaces in this way.

**Comment by thw0rted** [comment]
That's a good point about making it impossible to do a type-only import of the interface.  In my specific use-case I think that's probably not an issue but it's good to keep in mind.  My codebase is "very heavy on classes" (it's modern Angular, which is very class-centric) so I will probably just disable the rule on the assumption that my team probably won't use the "pseudo-module" antipattern anyway.

In future, would you consider a flag for the rule that treats namespaces which merge with a class (within the same file) differently from those that don't?  I recognize that it might not be a common enough pattern to warrant the extra work.

**Comment by bradzacher** [comment]
If you find it's a problem, we can consider it.

But in my experience - in modern TS people don't use namespaces at all because there's no need.
